velox/docs/functions/spark/regexp.rst

```diff
@@ -38,7 +38,9 @@ See https://github.com/google/re2/wiki/Syntax for more information.
 .. spark:function:: regexp_replace(string, pattern, overwrite) -> varchar
 
     Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite``. If no match is found, the original string is returned as is.
-    There is a limit to the number of unique regexes to be compiled per function call, which is 20.
+    There is a limit to the number of unique regexes to be compiled per function call, which is 20. If this limit is exceeded the function will throw an exception and return null values.
+    regexp_replace can also return null values if the any provided argument is null, ``string`` contains an invalid UTF-8 character, or ``pattern`` does not conform to RE2 syntax: https://github.com/google/re2/wiki/Syntax.
```

Comment:
"if the any" -> "if any"

> or ``pattern`` does not conform to RE2 syntax

Does Spark also return NULL if regexp is invalid?

Reply:
Shouldn't Velox implementation also throw in these cases?

Reply:
The only times this function returns 'null' is when an exception is thrown. Now that I am thinking about it, I originally believed the fuzzer errors I was running into were because I threw exceptions in a 'void call' function, but now that we figured out it was due to cache + dictionary encoding, I'm questioning my understanding of bool vs void call.

If the only times this function 'returns' null is after throwing an exception, would void call be the correct implementation? If that's the case, it would be a really quick fix to implement. I am currently building that version to test the fuzzer again

Reply:
@codyschierbeck 'void call' is appropriate if function either completes successfully and produces a non-null result or fails. 'bool call' is needed only if function may complete successfully and produce a null result.

Reply:
> Would "Optimize and re-regeister SparkSQL regexp_replace" be a better name for this PR?

That would be great. 

P.S. I'm so impressed with your ability to produce lengthly clear detailed writings in no time at all. Sorry that I cannot match this beautiful communication style in my terse responses.

Reply:
Thank you very much, that honestly made my day.

I will say that it is because I genuinely enjoy working on Velox. I concentrated my CS Master's in Big Data, but was never given an opportunity at work to leverage that until being assigned to Velox. So I was super excited to get started even though I knew practically zero C++ before diving head-first into this codebase. 

I'm probably not the best judge of this, but it is such an impressive codebase. It has been a fantastic in helping me learn. For example, today I was trying to learn how I might go about implementing either a Tuple or VariadicTuple type in case that is what the SpecialForm of the decode signature ends up requiring. My debugger ended up leading me into the UdfTypeResolver.h which made me look at Type.h, VELOX_FUNCTION_DEFINE_TYPES(T), and basically the entire framework in a whole new light.

That feeling happens pretty much every time I face a new problem with this codebase, so even though at times I've been pretty embarrassed for how obvious the solutions end up being, I enjoy it.